FK_solver/src/FK_solver.cpp

```diff
@@ -0,0 +1,341 @@
+                          *quadrature,
+                          update_values | update_gradients |
+                            update_quadrature_points | update_JxW_values);
+
+  FullMatrix<double> cell_matrix(dofs_per_cell, dofs_per_cell);
+  Vector<double>     cell_residual(dofs_per_cell);
+
+  std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+  jacobian_matrix = 0.0;
+  residual_vector = 0.0;
+
+  // Value and gradient of the solution on current cell.
+  std::vector<double>         solution_loc(n_q);
+  std::vector<Tensor<1, dim>> solution_gradient_loc(n_q);
+
+  // Value of the solution at previous timestep (un) on current cell.
+  std::vector<double> solution_old_loc(n_q);
+
+  forcing_term.set_time(time);
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (!cell->is_locally_owned())
+        continue;
+
+      fe_values.reinit(cell);
+
+      cell_matrix   = 0.0;
+      cell_residual = 0.0;
+
+      fe_values.get_function_values(solution, solution_loc);
+      fe_values.get_function_gradients(solution, solution_gradient_loc);
+      fe_values.get_function_values(solution_old, solution_old_loc);
+
+      for (unsigned int q = 0; q < n_q; ++q) {
+    // Evaluate coefficients on this quadrature node.
+    const double alpha_loc = alpha.value(fe_values.quadrature_point(q));
+
+    // Define the matrix to store D values and populate it.
+    FullMatrix<double> D_matrix(dim, dim);
+    D.matrix_value(fe_values.quadrature_point(q), D_matrix);
+
+    for (unsigned int i = 0; i < dofs_per_cell; ++i) {
+        for (unsigned int j = 0; j < dofs_per_cell; ++j) {
+
+            cell_matrix(i, j) += fe_values.shape_value(i, q) * 
+                                 fe_values.shape_value(j, q) / deltat * 
+                                 fe_values.JxW(q);
+
+            cell_matrix(i, j) -= alpha_loc * 
+                                fe_values.shape_value(j, q) * 
+                                fe_values.shape_value(i, q) * 
+                                fe_values.JxW(q);
+
+            cell_matrix(i, j) += 2.0 * alpha_loc *
+                                 solution_loc[q] * 
+                                 fe_values.shape_value(j, q) * 
+                                 fe_values.shape_value(i, q) * 
+                                 fe_values.JxW(q);
+
+            Tensor<1, dim> grad_phi_i = fe_values.shape_grad(i, q);
+            Tensor<1, dim> grad_phi_j = fe_values.shape_grad(j, q);
+            Tensor<1, dim> D_grad_phi_j;
+            
+            //compute the matrix vectol product
+            for (unsigned int d = 0; d < dim; ++d) {
+              //D_grad_phi_j[d] = 0.0;
+              for (unsigned int e = 0; e < dim; ++e) {
+                D_grad_phi_j[d] += D_matrix(d, e) * grad_phi_j[e];
+              }
+            }
+
+            cell_matrix(i, j) += scalar_product(D_grad_phi_j, grad_phi_i) *
+                                 fe_values.JxW(q);
```

Comment:
See my other comment on using `Tensor` to define the matrix $D$. All these lines would become
```cpp
cell_matrix(i, j) += scalar_product(D * grad_phi_j, grad_phi_i) * fe_values.JxW(q);
```
The only possible downside of this approach is that it requires the allocation of a temporary `Tensor<1, dim>` to store the result of `D * grad_phi_j`. However, this is usually an acceptable price to pay for the simplicity of the implementation (which entails readability and maintainability in the long run).

If (and only if) you should find out that this line becomes a major performance bottleneck, the following is a possible optimization:
```cpp
for (unsigned int c = 0; c < dim; ++c) {
  for (unsigned int d = 0; d < dim; ++d) {
    cell_matrix(i, j) += grad_phi_j[c] * D_matrix[c][d] * grad_phi_i[d];
  }
}
```
(where the syntax assumes that `D_matrix` is now a `Tensor<2, dim>`).